lib/xenops_server.ml

```diff
@@ -1235,11 +1235,11 @@ let rec perform_atomic ~progress_callback ?subtask ?result (op: atomic) (t: Xeno
 				 [ "-pci_passthrough"; String.concat "," sbdfs] in
 			B.VM.build t (VM_DB.read_exn id) vbds vifs vgpus extras force
 		| VM_shutdown_domain (id, reason, timeout) ->
+			debug "VM.shutdown_domain %s, reason = %s, timeout = %f" id (string_of_shutdown_request reason) timeout;
 			let start = Unix.gettimeofday () in
 			let vm = VM_DB.read_exn id in
-			(* Spend at most the first minute waiting for a clean shutdown ack. This allows
-			   us to abort early. *)
-			if not (B.VM.request_shutdown t vm reason (min 60. timeout))
```

Comment:
Was there a reason for having at least `60` seconds? Why not preserving the behaviour and allow only for longer timeouts?

Reply:
https://github.com/xapi-project/xen-api/commit/c51e4fbc4aa015364337442b4ca9181a7f7e8aa5 shows when this part was changed from using "at least 60 seconds" to "at most 60 seconds"; it doesn't explain why. However, the behaviour is preserved in this commit because we moved the use of `min <shutdown-timeout> <shutdown-ack-timeout>` from here to https://github.com/xapi-project/xenopsd/pull/381/files#diff-53e0cb0b01cbb6137c6b98996b38bd29R1375.

Reply:
If the user decides to change the default then that is their problem? Or do you mean that we should always keep a hard-coded lower limit for the ack timeout value?... I've seen no reason to think we do... @robhoes, @jonludlam?

Reply:
Oh I think I understand what you mean now; so you're suggesting to keep the use of `(min 60. timeout)` here but also keep the new configurable ack timeout in https://github.com/xapi-project/xenopsd/pull/381/files#diff-53e0cb0b01cbb6137c6b98996b38bd29R1375 ; that way we keep the current minimum 60 second ack timeout whilst having the ack timeout be configurable to longer periods... is this correct?

Reply:
Oh, yes, you are right, I had rewritten the `at most` into `at least` in my head, sorry. Still it may be worth clarifying it

Reply:
So the issue was that we couldn't make it longer than 60. seconds, so in this regard this change is good. What I am wondering is:

- should we have 60 second as minimum allowed timeout?
- is having a timeout longer than 60 second affecting other xenopsd timeouts?

Maybe the change from "at least" to "at most:" was related. I don't know.

Reply:
Regarding `min 60. timeout`: I think `timeout` is the total time we allow for the entire shutdown process, including the waiting for acknowledgement in the guest. There is a hardcoded 60s wait-for-ack timeout, _or less than 60s if the total shutdown time is less than that_. So all we need to do here is make this 60s configurable (again).

Reply:
So why did you change the meaning of the timeout param of `VM.request_shutdown` from "ack timeout" to "total timeout"? I think the former is more directly relevant to this function.

Reply:
@robhoes I agree that it would've been better to not change the param; my first solution was to just replace the `60.` with `!Xenopsd.domain_shutdown_ack_timeout` but it caused a circular dependency between `lib/xenopsd.ml` and `lib/xenops_server.ml`, so I can't reference `Xenopsd` from `Xenops_server`. I'm not sure if the design here is wrong or I am missing some workaround, but suggestions are welcome :)

Reply:
You can add the ref to resources.ml instead to get around the circular dependency.

Reply:
@robhoes thanks, that works!